parla/tasks.py

```diff
@@ -486,14 +486,27 @@ def reserve_persistent_memory(amount, device = None):
       be supplied if any element of the list is not an array. This may be a list of \
       devices if amount is a list of array objects.
     """
+    # TODO: This function should be split up into simpler subunits.
+    # How exactly that should be done isn't settled yet, but there's
+    # some discussion on this at
+    # https://github.com/ut-parla/Parla.py/pull/40#discussion_r608857593
+    # https://github.com/ut-parla/Parla.py/pull/40#discussion_r608853345
+    # TODO: reduce nesting by separating out the try/except idioms for
+    # checking if something supports the buffer protocol and checking
+    # whether or not something is iterable into separate functions.
+    # TODO: Generalize the naming/interface here to allow reserving
+    # resources other than memory.
     from . import cpu
     if isinstance(amount, int):
         memsize = amount
     elif hasattr(amount, '__cuda_array_interface__'):
         import cupy
         # cupy has the info we need and views are cheap,
         # so get a view then read off the size/device info.
-        view = cupy.array(amount)
+        view = cupy.array(amount, copy = False)
+        # TODO: there's actually a storage_size helper routine in
+        # parla.array. We should use it here and move the logic
+        # about buffer protocols there.
```

Comment:
The logic about buffer protocols goes in the buffer type implementations defined in `parla.cpu` and `parla.cuda`. `parla.array` has no concrete knowledge of how buffer types work.

Reply:
How do we want to consistently get the size of the buffers? Most things support itemsize/size/nbytes, but numba's DeviceNDArrays do not (https://numba.pydata.org/numba-doc/dev/cuda/memory.html#device-arrays). That's just one such example. There are lots of projects out there that use pep 3118/\_\_array_interface\_\_/\_\_cuda_array_interface\_\_ and our safest bet is to get memory use information through those kinds of protocols.

Reply:
I guess probably the takeaway is that this needs some more thought. Either way what we have is good enough for the time being.